src/parallel/mpi_concurrency/mpi_processor_grouping.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2018 ETH Zurich
+// Copyright (C) 2018 UT-Battelle, LLC
+// All rights reserved.
+//
+// See LICENSE.txt for terms of usage.
+// See CITATION.txt for citation guidelines if you use this code for scientific publications.
+//
+// Author: Giovanni Balduzzi (gbalduzz@itp.phys.ethz.ch)
+//
+// This file implements mpi_processor_grouping.hpp.
+
+#include "dca/parallel/mpi_concurrency/mpi_processor_grouping.hpp"
+
+#include <iostream>
+#include <unistd.h>
+
+#include "dca/parallel/mpi_concurrency/kernel_test.hpp"
+
+namespace dca {
+namespace parallel {
+// dca::parallel::
+
+MPIProcessorGrouping::MPIProcessorGrouping(bool (*check)()) {
+  // Initialize grouping with MPI world.
+  MPI_communication_ = MPI_COMM_WORLD;
+  MPI_Comm_size(MPI_COMM_WORLD, &world_size_);
+  MPI_Comm_rank(MPI_COMM_WORLD, &world_id_);
+
+  // Check if the process has the desired qualities, or remove it from the communicator.
+  const bool is_valid = check();
+
+  MPI_Comm_split(MPI_COMM_WORLD, static_cast<int>(is_valid), 0, &MPI_communication_);
+
+  MPI_Comm_size(MPI_communication_, &size_);
+
+  if (is_valid) {
+    MPI_Comm_rank(MPI_communication_, &id_);
+    if (size_ < world_size_)
+      printRemovedProcesses();
+  }
+  else {
+    id_ = -1;
+    printRemovedProcesses();
+  }
+}
+
+MPIProcessorGrouping::~MPIProcessorGrouping() {
+  MPI_Comm_free(&MPI_communication_);
+}
+
+bool MPIProcessorGrouping::defaultCheck() {
+#ifdef DCA_HAVE_CUDA
+  try {
+    return kernelTest();
+  }
+  catch (...) {
+    return false;
+  }
+#else
+  return true;
+#endif  // DCA_HAVE_CUDA
+}
+
+void MPIProcessorGrouping::printRemovedProcesses() const {
```

Comment:
Would it make sense to move this into `MPICocurrency` in case we want to keep communicators where not being part of it is expected behaviour?

Reply:
File an issue, we can consider additional refactoring after we're done this intergration.